cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go

```diff
@@ -122,6 +133,50 @@ type ClusterConfiguration struct {
 	ClusterName string `json:"clusterName,omitempty"`
 }
 
+// DNSAddOnType defines string identifying DNS add-on types
+type DNSAddOnType string
+
+const (
+	// CoreDNS add-on type
+	CoreDNS DNSAddOnType = "CoreDNS"
+
+	// KubeDNS add-on type
+	KubeDNS DNSAddOnType = "kube-dns"
+)
+
+// DNS defines the DNS addon that should be used in the cluster
+type DNS struct {
+	// Type defines string DNS add-on to be used
+	Type DNSAddOnType `json:"type"`
+
+	// ImageRepository, if set this fields allows to override the default image repository
+	// defined in ClusterConfiguration
+	ImageRepository string `json:"imageRepository,omitempty"`
+
+	// Image specifies which container image to use for running CoreDNS (this field does will be ignored if the selected
+	// add-on type is kube-dns).
+	// If empty, kubeadm will automatically select CoreDNS image using image repository and the supported CoreDNS version.
+	// In case this value is set, kubeadm does not change automatically the CoreDNS version during upgrades.
+	Image string `json:"image,omitempty"`
+
+	// ExtraArgs is a set of extra flags to pass to the dns add-on or override
+	// default ones in form of <flagname>=<value>
+	// TODO: Ideally we would like to switch all components to use ComponentConfig + ConfigMaps.
+	ExtraArgs map[string]string `json:"extraArgs,omitempty"`
```

Comment:
do you have anything in mind of how this is going to work?
our addons are defined in manifests that have API objects like deployments and daemon-sets.

"args" is ok for the control-plane because there we pass actual arguments to the containerized binaries, but in this case we are probably implying parameters to be replaced in a templated manifest?

can this bite us down the lane of having to maintain a sandbox of flags and such instead of letting the user provide custom manifests?

Reply:
I am supportive of this one. I do think, that we will have a lot more frustrated users, trying to roll on their own CoreDNS manifests just to add a single arg. This will make things better for them.

Reply:
i would call these flags instead of args.